### ISSUE TITLE
Operations not working with Microsoft.EntityFrameworkCore v2.1.0+

### ISSUE THREAD

**Issue by nikalfa** [json-api-dotnet/JsonApiDotNetCore#339]
Attempted to update from version 2.2.3 to the latest stable 2.3.3

As a consequence had to update Microsoft.EntityFrameworkCore.SqlServer and Microsoft.EntityFrameworkCore.Tools from 2.0.3 to 2.1.1

After that operations stopped working with the following error message:

```{
    "errors": [
        {
            "title": "InvalidOperationException",
            "detail": "Cannot consume scoped service 'JsonApiDotNetCore.Internal.Generics.IGenericProcessorFactory' from singleton 'JsonApiDotNetCore.Services.Operations.IOperationProcessorResolver'.",
            "status": "500",
            "meta": {
                "stackTrace": [
                    "System.InvalidOperationException: Cannot consume scoped service 'JsonApiDotNetCore.Internal.Generics.IGenericProcessorFactory' from singleton 'JsonApiDotNetCore.Services.Operations.IOperationProcessorResolver'.\r",
                    "   at Type Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitScoped(ScopedCallSite scopedCallSite, CallSiteValidatorState state)\r",
                    "   at TResult Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)\r",
                    "   at Type Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitConstructor(ConstructorCallSite constructorCallSite, CallSiteValidatorState state)\r",
                    "   at TResult Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)\r",
                    "   at Type Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitSingleton(SingletonCallSite singletonCallSite, CallSiteValidatorState state)\r",
                    "   at TResult Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)\r",
                    "   at Type Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitConstructor(ConstructorCallSite constructorCallSite, CallSiteValidatorState state)\r",
                    "   at TResult Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)\r",
                    "   at Type Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitScoped(ScopedCallSite scopedCallSite, CallSiteValidatorState state)\r",
                    "   at TResult Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)\r",
                    "   at void Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.ValidateCallSite(IServiceCallSite callSite)\r",
                    "   at Func<ServiceProviderEngineScope, object> Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.CreateServiceAccessor(Type serviceType)\r",
                    "   at TValue System.Collections.Concurrent.ConcurrentDictionary<TKey, TValue>.GetOrAdd(TKey key, Func<TKey, TValue> valueFactory)\r",
                    "   at object Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)\r",
                    "   at object Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetService(IServiceProvider sp, Type type, Type requiredBy, bool isDefaultParameterRequired)\r",
                    "   at object lambda_method(Closure, IServiceProvider, object[])\r",
                    "   at object Microsoft.AspNetCore.Mvc.Controllers.ControllerFactoryProvider+<>c__DisplayClass5_0.<CreateControllerFactory>g__CreateController|0(?)+CreateController(ControllerContext controllerContext)\r",
                    "   at Task Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)\r",
                    "   at async Task Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()\r",
                    "   at async Task Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextExceptionFilterAsync()"
                ]
            }
        }
    ]
}
```

Upon investigation I found out that updating either Microsoft.EntityFrameworkCore.SqlServer or Microsoft.EntityFrameworkCore.Tools to version 2.1.0+ brakes the operations - no matter the version of JAPDNC.

The dependency to SqlServer comes from this line in my startup:
 ```services.AddDbContext<DomainDbContext>(options => options.UseSqlServer(Configuration.GetConnectionString("DefaultConnection")), ServiceLifetime.Transient);```

**Comment by jaredcnance** [comment]
If you’re using operations, I highly recommend upgrading all the way to 2.4.0-beta1.

We are currently dogfooding this release which is why it is in beta but it has some essential fixes.

**Comment by nikalfa** [comment]
@jaredcnance I also tried updating to the beta release - the error persists.

Btw, I find the operations feature extremely useful, so I really do hope that this feature won't get dumped :)

**Comment by jaredcnance** [comment]
Operations are a high priority for us and are definitely not getting dumped 😄. I don't believe the error is actually related to the version of the SqlServer provider but the lib itself. It looks like it is caused by overwriting the service registration for `IGenericProcessorFactory`:

https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/404231bc2a96ee5cfbacf8b89c67b6af1d19a6eb/src/JsonApiDotNetCore/Extensions/IServiceCollectionExtensions.cs#L163

https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/404231bc2a96ee5cfbacf8b89c67b6af1d19a6eb/src/JsonApiDotNetCore/Extensions/IServiceCollectionExtensions.cs#L138

But really I think `IOperationProcessorResolver` and `IGenericProcessorFactory` should be scoped registrations. I'm not sure why this hasn't been caught in tests. I'll look into it further this afternoon, but you might be able to do something like this as quick fix:

```c#
services.AddJsonApi();
services.AddScoped<IOperationProcessorResolver, OperationProcessorResolver>();
```

**Comment by nikalfa** [comment]
This does indeed fix the issue. 

This issue actually might have originated a long time ago, since I did this in an old service, although from what I can remember only for debugging purposes.

**Comment by jaredcnance** [comment]
This has been fixed in v2.4.0-beta2.
